Generalized Pachinko

MrLongbottom opened the pull request:
Pachinko med 3 lag.

Commit:
day 1 progress

Commit:
day progress

switching up random initilization.
Chaning totoal counts into sparse matrixes showing per document topic counts.
Trying to fix math.

Commit:
finished rough slow implementation

Commit:
Update pachinko_gibbs_lda.py

Commit:
calculate sums for each document then update during sampling

rasmus-e-christensen committed:
Merge branch 'main' into feature/pachinko

Commit:
calculate topic-word sums out of loop and then update in loop

Commit:
Update pachinko_gibbs_lda.py

Commit:
accounting for observed data

Commit:
update todos

Commit:
cleanup pachinko main

rasmus-e-christensen committed:
added coherence measure to pachinko

Commit:
make pachinko work on any number of layers (WIP)

Commit:
topic is chosen based on flattened multinomial

Commit:
replace stack multiplication with einsum to multiply 2d matrices together into one 3d matrix

Commit:
change all multiplication in into einsum

Commit:
different way of finding argmax of 3d matrix

Commit:
now accounts for known taxonomy and skips calculations accordingly

Commit:
update iteration print

rasmus-e-christensen commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,205 @@
+    return word_topic_assignment, middle_layers, topic_to_word
+
+
+def decrease_counts(topics, middle_layers, middle_sums, topic_to_word, topic_to_word_sums, word, d):
+    for i, x in enumerate(middle_layers[d]):
+        x[topics[i], topics[i+1]] -= 1
+    for i, x in enumerate(middle_sums):
+        x[topics[i]] -= 1
+    topic_to_word[topics[len(topics)-1], word] -= 1
+    topic_to_word_sums[topics[len(topics)-1]] -= 1
+
+
+def increase_counts(topics, middle_layers, middle_sums, topic_to_word, topic_to_word_sums, word, d):
+    for i, x in enumerate(middle_layers[d]):
+        x[topics[i], topics[i + 1]] += 1
+    for i, x in enumerate(middle_sums):
+        x[topics[i]] += 1
+    topic_to_word[topics[len(topics) - 1], word] += 1
+    topic_to_word_sums[topics[len(topics) - 1]] += 1
```
Move this to gibbs ultility and call it pachinko_increase_counts and pachinko_decrease_counts

MrLongbottom replied:
hvis den er specifik kun til pachinko hvorfår så lægge den i en anden fil?

rasmus-e-christensen replied:
true

Commit:
added comments

Commit:
also show taxonomy name when printing topics

Commit:
Implemented support for multiple taxonomy chains

currently just chooses randomly between chains for each word

Commit:
small fix

Commit:
Pachinko now fully supports arbritrary number of layers

max layer size: 4
also supports both with and without 'lock' based on taxonomy tree

Commit:
minor fix

Seems I have swapped the order of two parameters for no reason.
Swapping back to make sure this doesn't influence anything else.

Commit:
Revert "minor fix"

This reverts commit cae763d28bcce7bfcb276df4db22139bbc26056c.

rasmus-e-christensen commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,267 @@
+        if len(tax) == 1 and len(tax[0]) == len(layer_lengths):
+            continue
+
+        # sims calculated to be used later
```
```suggestion
        # sums calculated to be used later
```

rasmus-e-christensen commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,267 @@
+            if len(word_tax) < len(layer_lengths):
+                divs.append(np.divide((topic_to_word[:, word] + beta), topic_to_word_sums + (M * beta)))
+
+            # TODO make work for any number of dimensions (currently only working with 3 middle layers)
```
er dette gjort?

MrLongbottom replied:
ja
```suggestion
```

Commit:
Update model/pachinko_gibbs_lda.py

Co-authored-by: Rasmus Christensen <rech16@student.aau.dk>

Commit:
Update model/pachinko_gibbs_lda.py

Starrock45 commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def gibbs_sampling(documents: List[np.ndarray],
+                   wta,
```
Vil nok hellere have det skrevet ud, da det alligevel kun bruges 1 sted i koden

MrLongbottom replied:
altså i main? så bliver main nok meget lang, men det kan vi godt?

Starrock45 replied:
```suggestion
                   word_topic_assignment,
```
Jeg tænkte sådan her, og så ændre den på linje 113. Linje 174 bruger nemlig også lige nu word_topic_assignment fra main.

MrLongbottom replied:
right forstår hvad du mener nu.

Starrock45 commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,266 @@
+                   topic_to_word):
+    """
+    Takes a set of documents and samples a new topic for each word within each document.
+    :param middle_layers:
```
Det ville være dejligt hvis `middle_layers` også blev beskrevet.

MrLongbottom replied:
har opdateret til bare at bruge de variabler som er defineret i main, frem for at tage dem alle med som parametre

Starrock45 commented on model/pachinko_gibbs_lda.py:
```diff
@@ -0,0 +1,266 @@
+            topic = tuple(topic)
+
+            # Assign and increase topic count
+            word_topic_assignment[d_index][w_index] = topic
```
Er nok mere passende at bruge `wta` fra parametrene, medmindre parameternavnet alligevel ændres.

Starrock45 reviewed (commented):
Har et par kommentarer. Jeg har nok ikke forstået det hele fuldstændigt, men jeg har heller ikke læst pachinko artiklen endnu.

Commit:
minor fixes

Commit:
Merge branch 'main' into feature/pachinko_general

Commit:
redo coherence

hopefully this will make sure that other code that uses this function doesn't fuck up.

Commit:
merge
